re-try tar command if we encounter an error

npalaska opened the pull request:
Part of an issue https://github.com/distributed-system-analysis/pbench/issues/2692 
Addresses Peter's comment about retrying the tar command if we encounter an error for the first time
> Let's ignore all warnings on the re-try only. Something like `tar --warning=none`, so that we capture the errors when we try the first time, but the tar up whatever we can on the re-try. And we should do something to remark in the data collection that we encountered warnings on the initial `tar` command.

dbutenhof commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1316,121 +1316,120 @@ def _send_directory(self, directory, uri, ctx):
-                    failures += 1
-                else:
+                    # record the tar error we encountered first time
+                    with open(target_dir / "tar_command_error") as file:
+                        file.write(str(cp.stdout))
```
Or `(target_dir / "tar_command_error").write_text(str(cp.stdout))`??

portante replied:
Or why not use the Redis logger?

portante replied:
The more I think about it, there is nothing that will consume this file in the tar ball to take a different action.  For now having the warning log message above is sufficient.

dbutenhof commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1316,121 +1316,120 @@ def _send_directory(self, directory, uri, ctx):
+                            stdout=subprocess.PIPE,
+                            stderr=subprocess.STDOUT,
+                        )
+                        assert (
+                            cp.returncode == 0
+                        ), "Failed to create tar file second time"
                     except Exception:
```
The assert seems unnecessarily harsh; for one thing, we won't be logging the exception and incrementing `failures` ... or ... well, actually we *will*, because your `except Exception` will immediately intercept the `assert` 😄 

Also, again, the point of an `assert` is that it sublimates without residue in a "production" `-O` build, and we'll still want to recognize and handle failure of the retry in that case.

npalaska replied:
Yeah, the command should theoretically succeed because we specify `--warning=none` , so this assertions is to just reiterate that we got `returncode=0` with this option and if any exception occurs we still have the exception block to catch it. If the returncode is not 0 for any reason for the 2nd time even after specifying `--warning=none` we should exit isn't it? and failed assertion will do that for us.

dbutenhof replied:
Look at the call sites:

```
        failures = self._send_directory(tool_dir, "tool-data", tool_data_ctx)

        if failures == 0:
            del self.directories[directory]

        self._send_client_status(
            "success" if failures == 0 else f"{failures} failures sending tool data"
        )
```

If something goes wrong with the `tar` and we `assert`, we'll never send the failure message to the TDS, and the whole thing hangs up because TDS never got all the send confirmations. That would be bad.

webbnh replied:
Peter illuminated this for me:  his request for this PR comes from a scenario which Lukas Doktor hit where Pbench was trying to capture the run results while SAR was still writing them, which caused the files to change while they were being archived, which `tar` gets upset about.  So, the point of the retry is to overcome `tar`'s concerns on that front.

I believe that Peter recommended that we not concern ourselves with other modes of failure, here (which, perhaps, is a good idea).  That being the case, I would recommend tailoring the error handling here toward the "files changed" failure that Lukas hit, to the extent that it is possible -- that seems like a realistic and specific problem scenario, and it seems like one that we might address without "boiling the ocean"....  (And, it's something that we could explain in the code....)

npalaska replied:
> I would recommend tailoring the error handling here toward the "files changed" failure that Lukas hit, to the extent that it is possible

So you mean instead of doing `warning=None` we do `warning=file-changed --ignore-failed-read` that should suppress the warnings about changing files while we are archiving

dbutenhof replied:
We'll never be bullet-proof here, so that suggestion has some merit. If there are *particular* default warnings we know we can successfully ignore to have a better chance of getting a dataset (and aren't 100% sure we understand the consequences of ignoring *all* warnings... which I'm certainly not), then picking the ones we know may make sense. With that said, I don't object to `--warning=none` as a generic "let's just do our best to get some artifact we can bring back to the museum".

webbnh replied:
My intention was to change to react only to specific modes of failure, but that seems to be prohibitively difficult (i.e., it requires parsing `stderr`, which seems like a bad idea).

Once we decide to retry, I think we should grab whatever we can get and flee (although, I am slightly worried that by the time we get back to the museum we will have forgotten that we left in a panic...).

portante commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1316,121 +1316,120 @@ def _send_directory(self, directory, uri, ctx):
+            "--force-local",
+            f"--file={tar_file}",
+            target_dir,
+        ]
```
Perhaps it might be better to create a method which runs the tar command, and then re-runs it so that this caller can just invoke that with the details of that process hidden?

portante reviewed (changes requested):
Can we consider having a separate function/method for running the tar commands to encapsulate the process of re-trying?

Also, would like to see the Redis logger used for the tar command output.

portante commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1316,121 +1316,120 @@ def _send_directory(self, directory, uri, ctx):
-                        "Failed to create tar ball; return code: %d", cp.returncode,
+                    self.logger.warning(
+                        "Failed to create tar ball first time; Error: %s, re-trying with --warning=none",
+                        str(cp.stdout),
```
Perhaps consider using `cp.stdout.decode("utf-8")`.

Also, perhaps re-wording to consider?  `Tar ball creation failed with '%s', re-trying ignoring all warnings using --warning=none`

dbutenhof replied:
I wondered if we should include `cp.returncode`? Looking at `tar` and `xz` I'm not sure these are particularly valuable (basically it'll be 1 or 2, and interpreting the actual meaning will be down to analyzing the `stderr` output anyway). Still... perhaps we should?

npalaska replied:
Yep adding returncode doesn't hurt, added it in the latest commit.

webbnh reviewed (changes requested):
After talking with Peter, I have the following proposal.

First, `_send_directory()` is way too complicated...that's not your fault, Nikhil, but adding more in-line code to it should be avoided.  Therefore, please take Peter's suggestion to encapsulate your new code in a subroutine.

Second, this PR doesn't include any unit test changes or additions, which it probably should:  you've added new code and new functionality, and that should come with unit-testing.  However, if you provide your new code in a subroutine, then it should be straightforward to include tests which exercise it.

Finally, the existing unit tests for `_send_directory()` need to be updated to account for your new code.  In particular, if your code comes in in a subroutine with unit tests, then the existing unit testing does not need to be responsible for testing your new code.  Instead, you should add mocking to the existing unit tests to drive whatever behavior they require vis a vis your new code.  (The details of that will depend on how your new code is packaged...it's possible that the existing code won't require much at all.)

If you have any questions, please let me know.

webbnh commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1316,121 +1316,120 @@ def _send_directory(self, directory, uri, ctx):
-                    cwd=parent_dir,
-                    stdin=None,
-                    stdout=ofp,
-                    stderr=efp,
-                )
+            cp = subprocess.run(
+                tar_args,
+                cwd=parent_dir,
+                stdin=None,
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+            )
```
Perhaps not in this PR...but I can dream....  😿

npalaska replied:
Well we still need a `try` block because we are running the `tar` command directly and if the `tar_path` is somehow got corrupt it would result in an exception.

webbnh replied:
True, if `self.tar_path` were to go bad between the call to `find_executable()` and here, we would get an exception raised by `run()`.  However, I would contend that that is so unlikely that it's not worth the cost and complication of trying to catch it locally -- I think that that's tantamount to a catastrophic failure, and we should deal with that at the top-most level if at all.

npalaska commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1316,121 +1316,120 @@ def _send_directory(self, directory, uri, ctx):
+                        )
                         failures += 1
-                    else:
-                        try:
-                            o_file.unlink()
-                        except Exception as exc:
-                            self.logger.warning(
-                                "Failure removing tar command output file, %s: %s",
-                                o_file,
-                                exc,
-                            )
+                try:
+                    (_, tar_md5) = md5sum(tar_file)
```
Yeah it will hang, so if we failed the tar creation on the 2nd attempt, Peter suggested me that we at least need to send an empty tar file so that it wont hang.

webbnh replied:
We need to make sure that we produce a file which we can successfully hash so that we end up performing the send....

dbutenhof replied:
Of course if the drive is full, even an "empty" tar file takes up space and might not succeed. In which case we need to be prepared for failure.

I'd feel better about this path if we could say "we made two attempts and it didn't work, so tell the TDS control channel that we can't generate an artifact and skip the `PUT` entirely". (If we can't even do *that*, there's no hope.)

webbnh commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1285,6 +1285,50 @@ def stop_tools(self, data):
+            if cp.returncode != 0:
+                self.logger.warning(
+                    "Failed to create tar ball with arguments %s; Error: %s",
+                    str(tar_args),
```
Elsewhere, instead of stringifying a list of strings, we've been `join`'ing them to good effect:
```
                    " ".join(tar_args),
```

webbnh commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1285,6 +1285,50 @@ def stop_tools(self, data):
+                stdout=subprocess.PIPE,
+                stderr=subprocess.STDOUT,
+            )
+            if cp.returncode != 0:
+                self.logger.warning(
+                    "Failed to create tar ball with arguments %s; Error: %s",
+                    str(tar_args),
+                    cp.stdout.decode("utf-8"),
+                )
+        except Exception:
+            self.logger.exception(
+                "Failed to create tar ball with arguments '%s'", str(tar_args)
+            )
+        else:
+            return cp
+        return None
```
I don't know how everyone feels about this, but my preferences are to minimize the code inside the `try` block and to return immediately when we fail.  So, I would write this blob like:
```
        except Exception:
            self.logger.exception(
                "Failed to create tar ball with arguments '%s'", " ".join(tar_args)
            )
            return None

        if cp.returncode != 0:
            self.logger.warning(
                "Failed to create tar ball with arguments %s; Error: %s",
                " ".join(tar_args),
                cp.stdout.decode("utf-8"),
            )

        return cp
```

dbutenhof replied:
This is the client side of a distributed system. Yeah, we always have to deal with the possibility of total nuclear disaster with timeouts and such, but we shouldn't ever just casually drop out without trying to maintain the protocol. We should at least stick a message in a bottle and toss it overboard as we sink ...

webbnh commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1311,125 +1355,83 @@ def _send_directory(self, directory, uri, ctx):
+            # Re-try creating tar ball by suppressing all the warnings
+            tar_process = self._create_tar(target_dir, tar_file, ignore_warning=True)
+            if not tar_process:
+                # Tar ball creation failed even after suppressing all the warnings,
+                # we will now proceed to send an empty tar ball
+                self._create_tar(
+                    Path("/dev/null"), tar_file, cwd=parent_dir, ignore_warning=True
                 )
```
Are you sure you want to go to the trouble of trying to create an empty archive, rather than just creating a completely empty file?  (I mean, at this point, the `tar` command has failed twice already...it might fail a third time....)

If this third call fails, will we hang?

npalaska replied:
> If this third call fails, will we hang?

Yes, if this fails then the TDS hangs 

> Are you sure you want to go to the trouble of trying to create an empty archive, rather than just creating a completely empty file? (I mean, at this point, the `tar` command has failed twice already...it might fail a third time....)

The `put_document` of the tool data sink write this file as `tar.xz` on the controller, can we just send a regular file over to the tds and will the tds write it as a tar file or will that produce an exception? 🤔

dbutenhof replied:
```
[root@dbutenho-dev ~]# touch foo.tar.xz
[root@dbutenho-dev ~]# tar xvf foo.tar.xz 
tar: This does not look like a tar archive
xz: (stdin): File format not recognized
tar: Child returned status 1
tar: Error is not recoverable: exiting now
[root@dbutenho-dev ~]# echo $?
2
```

So, yeah; the `PUT` will run into a problem unpacking the file, causing `PUT` to terminate with a 500 status ("Failed to extract tar ball"), which will cause `_send_directory` to log a message and count a failure. Looks like the `PUT` will *not* update `_tm_tracking[hostname]` from "waiting" to "dormant", so the TDS will continue waiting for an artifact, and we haven't affected the hang.

At least passing the empty (real) tarball means that TDS can unpack and track receipt. Arguably, when TDS gets an upload that it can't unpack it should mark that host "failed" in the tracker and we should stop waiting. But, again, it'd be better to be able to skip the `PUT` entirely in this case and simply Redis a failure (yes, I've verbified that noun) that the TDS wait cycle will recognize as "no soup for you" and move on.

npalaska replied:
> Arguably, when TDS gets an upload that it can't unpack it should mark that host "failed" in the tracker and we should stop waiting

Yeah ideally we would want to do that, but its bit tricky to do it outside the tds because there is a watcher thread in tds that change the tracking status to dormant after the put finishes. I will put a TODO comment in our code to address this thing separately. This will probably involve changes to how TM and TDS communicate with each other.

dbutenhof replied:
If it sounded like I was trying to push you into fixing that now, I apologize; what I was intending was more to justify your temporary workaround... with acknowledgement that there's a better way. Peter already entered #2709 to approach this problem.

webbnh reviewed (commented):
Looks good!

I've made a couple of suggestions.

Looking forward to the new unit tests. 😉

npalaska commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1285,6 +1285,42 @@ def stop_tools(self, data):
             self._send_client_status("success")
         return failures
 
+    def _create_tar(
+        self,
```
If it make sense, should this be made a `utils` functionality and should this reside in `lib/pbench/common/utils.py` just like how we do it with `md5sum`?

dbutenhof replied:
Right now the server has no interest in creating a tarball, although we might eventually come up with a use case; so `agent/utils.py` might make more sense. Then again, the only other place I think we do this, where we're preparing to push data up to a server, is currently in `bash`.

dbutenhof commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1285,6 +1285,42 @@ def stop_tools(self, data):
+        self,
+        directory: Path,
+        tar_file: Path,
+        ignore_warning: list = [],
```
Semantically, the name `ignore_warning` is misleading; we're directly supplying keywords to the `--warning` option which *enables* warnings. We happen to be enabling `none`, which equates to disabling all, but since you're providing the option of making it a list, some caller might supply `["file-shrank", "no-failed-read"]`.

Without any specific use case for specifying specific warnings here, I might be tempted to go for a simple `ignore_warnings: bool = False`. Or, alternatively, `warning_keywords: List[str] = []`

And adding Args and Returns to the docstring would be a good idea to clarify what the parameters mean.

npalaska replied:
Yes, I tried to address this with some refactoring.

dbutenhof commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1311,125 +1347,90 @@ def _send_directory(self, directory, uri, ctx):
+                if tar_process.returncode != 0:
+                    # Tar ball creation failed even after suppressing all the warnings,
+                    # we will now proceed to create an empty tar ball
+                    self._create_tar(
+                        Path("/dev/null"), tar_file, ignore_warning=["none"]
+                    )
```
This empty tar file thing is weird. It occurs to me that, if the protocol really requires something to be `PUT` to TDS to continue (it'd be better to just be able to tell the TDS listener that we have nothing to PUT), maybe we should consider storing the tar status and `stderr` stream in a file in the directory so it gets recorded for triage as part of the run?

npalaska replied:
Yeah but we already log through the Redis, so the error message should already be there on the controller in tm.log file

dbutenhof reviewed (commented):
Did you amend a commit? Some of the comment chains I can't find in either of the two commits; though perhaps they were only made in the "Conversation" tab in the first place?

dbutenhof reviewed (commented):
I thought I posted these this morning, but apparently I rushed off for scrum and left it. I'm not entirely sure I finished, but I'm going to publish what I had and then look again.

webbnh commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1349,19 +1357,13 @@ def _send_directory(self, directory, uri, ctx):
+                # Tar ball creation failed even after suppressing all the warnings,
+                # we will now proceed to create an empty tar ball.
+                # TODO: it'd be better to be able to skip the PUT entirely if the
+                #  tar fails and simply log a failure without TDS waiting forever
```
I think there are a few leading spaces **_missing_** from this line.  🤣 

@dbutenhof, I think, if he omits that leading space, then his IDE won't treat this line as a continuation of the `TODO` comment.  (PyCharm has special handling for `TODO`'s and `FIXME`'s.  😁)

webbnh commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1349,19 +1357,13 @@ def _send_directory(self, directory, uri, ctx):
+                # we will now proceed to create an empty tar ball.
+                # TODO: it'd be better to be able to skip the PUT entirely if the
+                #  tar fails and simply log a failure without TDS waiting forever
+                if self._create_tar(Path("/dev/null"), tar_file, retry=True) != 0:
+                    raise ToolMeisterError
```
I still think that our fallback here should be an empty _file_, not an empty _`tar` archive_.  (I.e., we can create an empty file without having to run the `tar` command, and that saves us from needing a _fourth_ fallback position.)

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,113 @@
+        redis_server=None,
+        logger=None,
+    )
+    tmp_dir = agent_setup["tmp"]
+
+    # create a file in tmp directory
+    (tmp_dir / "file1").write_text("")
```
Assuming that we are mocking out the subroutines that `_create_tar()` calls, we shouldn't actually need a temp directory or any file system artifacts.

On the flip side, I don't see the mock for `subprocess.run()` here....

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,113 @@
+    cp = tm._create_tar(tmp_dir, tar_file)
+    assert cp.returncode == 0
+    assert cp.stdout == b""
+    tar_file.unlink()
```
We shouldn't be letting `_create_tar()` actually create a tar file, so there shouldn't be anything to unlink.

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,113 @@
+
+    cp = tm._create_tar(tmp_dir, tar_file)
+    assert cp.returncode == 1
+    assert b"file changed as we read it" in cp.stdout
```
We should be mocking `subprocess.run()` (not `tar`), and the mock should probably validate its arguments as well as returning the appropriate `stderr` and exit status.

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,113 @@
+    tar_file.unlink()
+
+
+def test_create_empty_tar(agent_setup, get_params):
+    """Test empty tar creation"""
```
This test isn't really useful:  we don't need to confirm that `tar` works; what we need to confirm is that the CUT is invoking it properly (which we can do with a mock).

(The _functional_ testing can concern itself with whether we create the tarball correctly -- the unit test is concerned with the execution of the code.)

webbnh reviewed (changes requested):
So, this is a good first draft, but it needs some work:
- `_create_tar()` should probably be split into two functions:
  - one which basically just invokes `subprocess.run()` in order to run `tar`
  - one which takes the directory and archive to create and encapsulates the retry logic, allowing it to reuse the code in the second.  (The recursive approach was clever, but it ends up being too cumbersome.)
- The unit tests should be testing only the new code
  - They shouldn't be creating, examining, or removing file system artifacts, because they should be mocking out the call to `subprocess.run()`, such that `tar` never actually executes.
  - Roughly speaking, there should be two unit test cases for every `if` statement in the CUT (and, if the `if` statements are nested, then that's three or four test cases).  But, after restructuring, the CUT will only have 2-4 `if` statements, spread across two functions, so we'll only need four, maybe six, tests.
  - It would be nice to have a couple of test cases for `_send_directory()` -- just ones which drive execution to the call to `_create_tar()`, to test how the new code there responds to the success and failure cases which it returns.  (And, for these, **do** mock out `_create_tar()`!)

(Most of my comments are on the code as it stands, so they won't really apply if you restructure things...but, they are included for completeness.)

webbnh commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1285,6 +1285,50 @@ def stop_tools(self, data):
+                "re-trying ignoring all warnings using --warning=none" if retry else "",
+                cp.returncode,
+            )
+            if retry:
+                tar_args.insert(2, "--warning=none")
+                self._create_tar(directory, tar_file, tar_args)
+            else:
+                return cp
+        else:
+            return cp
```
_[Hopefully you will restructure this code and this comment will be moot, but....]_

In the event that we make the recursive call to `_create_tar()` at line 1326, I think we'll return `None` instead of a `CompletedProcess` object.

In any case, I don't think we need multiple `return` statements, here -- I think you can remove all the `else`'s and just return `cp` at the end (assuming that we set it properly...).

webbnh commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1311,125 +1355,84 @@ def _send_directory(self, directory, uri, ctx):
+                # we will now proceed to create an empty tar ball.
+                # TODO: it'd be better to be able to skip the PUT entirely if the
+                #  tar fails and simply log a failure without TDS waiting forever
+                if self._create_tar(Path("/dev/null"), tar_file, retry=True) != 0:
+                    raise ToolMeisterError
```
You can add message text to the `ToolMeisterError`, and I would recommend doing so.

I would also recommend adding a comment here which points out that, when we raise this exception, we're going to skip the `PUT`, below, and that's going to cause the TDS to hang....

webbnh commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1289,21 +1289,22 @@ def _create_tar(
         directory: Path,
         tar_file: Path,
-        ignore_warning: list = [],
+        tar_args: List = [],
```
FWIW, this is what [the LGTM bot is complaining about](https://github.com/distributed-system-analysis/pbench/pull/2704#issuecomment-1077597891) (although, apparently it's not smart enough to determine that the problem cannot actually happen...or I'm wrong and it's smarter than I am... 😆).

dbutenhof commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1289,45 +1289,45 @@ def _create_tar(
-                "re-trying ignoring all warnings using --warning=none" if retry else "",
                 cp.returncode,
+                directory,
+                "re-trying ignoring all warnings using --warning=none" if retry else "",
```
You should have included the trailing "; " as well, because right now the non-`retry` case will look weird. However, I think it'd be better to just drop this entire phrase anyway. If you want, just include the value of the `retry` parameter.

npalaska replied:
For non-retry cases it looks like this:
`
__logger__ _create_tar 1320 -- Tar ball creation failed with 'Some error running tar' and returncode: 1, on directory /tmp/pytest-of-nikhilpalaskar/pytest-193/agent-tmp0; 
`

and for the retry case it looks like this 
`__logger__ _create_tar 1320 -- Tar ball creation failed with 'Some error running tar' and returncode: 1, on directory /tmp/pytest-of-nikhilpalaskar/pytest-192/agent-tmp0; re-trying ignoring all warnings using --warning=none` 

I dont think hey look weird, its just verbose warning message

webbnh replied:
What looks weird is the trailing semicolon at the end of the message.  Messages normally end with a period (or exclamation mark, or question mark...), if they end with any punctuation at all; in English text, a semicolon implies that there is another independent clause coming....

dbutenhof commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -2,15 +2,59 @@
+@pytest.fixture()
+def mock_tar(monkeypatch):
+    def fake_run(*args, **kwargs):
+        def f():
+            return
+
+        f.returncode = 0
+        f.stdout = b""
+        return f
+
+    monkeypatch.setattr(subprocess, "run", fake_run)
```
Huh? When you call `subprocess.run` you get back a *function* with `returncode` and `stdout` properties? I don't understand the logic here.

I'd suggest returning a `CompletedProcess` (or at least something that quacks like a `CompletedProcess` allowing the caller to access the necessary properties.

npalaska replied:
> Huh? When you call `subprocess.run` you get back a _function_ with `returncode` and `stdout` properties? I don't understand the logic here.

Yes the returned function has `returncode` and `stdout` attributes attached to it.

> I'd suggest returning a `CompletedProcess` (or at least something that quacks like a `CompletedProcess` allowing the caller to access the necessary properties.

True, I can return something like `subprocess.CompletedProcess(args, returncode=0, stdout=b"", stderr=None)`

npalaska commented:
I wonder what we need to do with all these warning messages getting printed on the console, We are testing them and they are supposed to be raised, but they are annoying when we look at the console outputs, should we be suppressing them? 

```
..2022-03-25 21:08:55,602 WARNING 313 140593309239104 __logger__ _create_tar 1320 -- Tar ball creation failed with 'Some error running tar' and returncode: 1, on directory /var/tmp/jenkins/tox/py/tmp/popen-gw12/agent-tmp0; 
..2022-03-25 21:08:55,605 WARNING 322 140278485026624 __logger__ _create_tar 1320 -- Tar ball creation failed with 'Some error running tar' and returncode: 1, on directory /var/tmp/jenkins/tox/py/tmp/popen-gw14/agent-tmp0/test.hostname.com; re-trying with --warning=none
2022-03-25 21:08:55,605 WARNING 322 140278485026624 __logger__ _send_directory 1362 -- Failed to create tarball, b'Some error running tar'
2022-03-25 21:08:55,606 WARNING 322 140278485026624 __logger__ _create_tar 1320 -- Tar ball creation failed with 'Some error running tar' and returncode: 1, on directory /dev/null; 
2022-03-25 21:08:55,607 ERROR 322 140278485026624 __logger__ _send_directory 1372 -- Failed to create tar ball, '/var/tmp/jenkins/tox/py/tmp/popen-gw14/agent-tmp0/test.hostname.com.tar.xz'
Traceback (most recent call last):
  File "/var/tmp/jenkins/tox/py/lib/python3.9/site-packages/pbench/agent/tool_meister.py", line 1370, in _send_directory
    raise ToolMeisterError("Failed to create an empty tar")
```

dbutenhof commented:
> I wonder what we need to do with all these warning messages getting printed on the console, We are testing them and they are supposed to be raised, but they are annoying when we look at the console outputs, should we be suppressing them?

Normally, `pytest` doesn't show the log messages unless the test case fails. Maybe it's showing `WARNING` and `ERROR` anyway? That's odd since you're supposed to be testing error conditions.

You could be using the `caplog` fixture to check that you're *getting* the warnings and errors you expect; I wonder if that would convince it to stop writing them out during the test run? Weird...

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -2,15 +2,49 @@
+    def fake_run(*args, **kwargs):
+        return subprocess.CompletedProcess(args, returncode=0, stdout=b"", stderr=None)
+
+    monkeypatch.setattr(subprocess, "run", fake_run)
```
You should consider having the `run()` mock assert:
- that the `cwd` keyword-arg is the `directory` parent
- that the `stdin`, `stdout`, and `stderr` kwargs have the expected values
- and that `tar_args` contains
  - `"--create"`
  - `"--xz"`
  - `"--force-local"`
  - the right output file
  - the right input path
  - and a reasonable-looking value for the tar executable path

I believe that fixtures can accept parameters, so you can pass in the values for the input, output, and current directories (or you can make them global symbols, since you use the same values in each test); the inner function can access them as a closure.  The rest of the validation can be delegated to a helper function, so that you can share it across the three fixtures.

You can also consider making the mock function "stateful", so that it knows that it should _not_ receive the `--warning=none` on the first call and that it _should_ receive it on the second call.  (I believe that if you store the state in the fixture, outside the inner function, then it will be modifiable and "sticky" in the inner function.)

You might also want to consider another parameter to the fixture which governs whether it succeeds or fails on the first or second try.  The downside of this is that it makes the code more complex; the upside is that there are fewer mock functions.

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -22,92 +56,98 @@ def get_params():
+            tmp_dir=None,
+            tar_path="tar_path",
+            sysinfo_dump=None,
+            params=TestCreateTar.tm_params,
```
I believe that you can access `tm_params` via `self`, which seems a little weird at first, but it's nicer because it means that, if you change the name of the class for some reason, you don't have to update the reference.

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -22,92 +56,98 @@ def get_params():
+            redis_server=None,
+            logger=None,
+        )
+        tmp_dir = agent_setup["tmp"]
+        tar_file = "test.tar.xz"
+
+        cp = tm._create_tar(tmp_dir, pathlib.Path(tar_file))
```
Since we're mocking out the operation, the `tmp_dir` doesn't actually have to exist...and, in fact, it's probably better if it _doesn't_, so I would use some other value here.  (Ditto for the other tests.)

npalaska replied:
Addressed.

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -22,92 +56,98 @@ def get_params():
+
+        cp = tm._create_tar(tmp_dir, pathlib.Path(tar_file))
+        assert cp.returncode == 1
+        assert cp.stdout == b"Some error running tar"
```
It would be better if there were a common definition for what the mock returns and for what we test against here (i.e., at the module level, define a symbol for the error message string, and then reference that symbol in the mocks and here).  Also, it would be good if the messages from the first try and the retry were different (if you make them so that they can both fail), so that we can ensure that we're seeing the correct one.

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -22,92 +56,98 @@ def get_params():
+class TestSendDirectory:
+    """Test create_tar in send directory of tool meister"""
+
+    def test_tar_create_success(self, agent_setup, mock_tar, caplog):
+        """This test should pass the tar creation in send directory"""
+
+        tm = ToolMeister(
+            pbench_install_dir=None,
+            tmp_dir=None,
+            tar_path="tar_path",
+            sysinfo_dump=None,
+            params=TestCreateTar.tm_params,
+            redis_server=None,
+            logger=logger,
+        )
+        directory = agent_setup["tmp"] / f"{TestCreateTar.tm_params['hostname']}"
+        failures = tm._send_directory(directory, "", "")
+        assert "Failed to create an empty tar" not in caplog.text
+        assert failures == 1
```
Why are we expecting a failure?  This should have a big ole code comment explaining it...either that, or you should be mocking `md5sum()`, `shutil.rmtree()`, `Path.unlink()`, etc., as well as using the `Requests` mocking for `PUT`, to drive this case to an actual successful return.

(I realize that testing `_send_directory()` is bear, but this test is nominally a "success case", so we need to make it clear that it is not.)

Also, for this test and the next one, you should be mocking out `_create_tar()` -- that is, the previous unit tests should be sufficient to test that function's code -- what we want to be testing here is the code which _calls_ the new function.  The mock should be asserting that the arguments it receives are appropriate, and it should generate the appropriate return value directly, without relying on mocking `subprocess.run()` to do it.  (I.e., the "unit" here should be `_send_directory()`, not "`_send_directory()` plus everything it calls".)

npalaska replied:
Yeah I had to add quite a lot of mocks for `send_directory` to succeed.

webbnh reviewed (approved):
The use of the inner function is much nicer than the recursive approach.

I have the usual pile of nits and suggestions for you to consider.  I would encourage you to make more use of the mocks for assertions.

Also, I would encourage you to mock out `_create_tar()` instead of `subprocess.run()` in the `_send_directory()` tests -- there's no need to include `_create_tar()` in the `_subprocess_run()` "unit".

portante commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,153 @@
+
+from pbench.agent.tool_meister import ToolMeister, get_logger
+
+logger = get_logger("__logger__")
```
I don't think a global logger should be used.  That creates state carried between tests.

portante replied:
Still applies.

portante commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,153 @@
+"""
+Tests for the Tool Data Meister modules.
```
Technically it is called the `Tool Meister` module.

portante commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1311,125 +1355,88 @@ def _send_directory(self, directory, uri, ctx):
+                # tar fails and simply log a failure without TDS waiting forever.
+                if self._create_tar(Path("/dev/null"), tar_file) != 0:
+                    # Empty tarball creation failed, so we're going to skip the PUT
+                    # operation, and that's going to cause the TDS to hang.
```
It is not clear to me we need to describe what the Tool Data Sink is going to do here.  For this comment, I would simply state that in this case the Tool Meister cannot hold up its end of the contract expected by the Tool Data Sink.

portante reviewed (commented):
Looking good ...

webbnh commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1311,125 +1355,88 @@ def _send_directory(self, directory, uri, ctx):
-                )
+            cp = self._create_tar(directory, tar_file, retry=True)
+            if cp.returncode != 0:
+                self.logger.warning("Failed to create tarball, %s", cp.stdout)
```
I went back and forth on this, @dbutenhof, and I ultimately decided that the callee should report the failure to the caller and it should be up to the caller to report the failure to the log.  That is, in some abstract sense, if the callee reports it, then the caller can't prevent that; and, worse, the caller doesn't know whether it's been logged or not, so the caller logs it and now it's been logged twice.

So, I settled on having the callee log stuff which is beyond the callers's control (i.e., the _first_ failure **when** its been instructed to retry) and then report the result (of the last attempt, either the only attempt or the retry) to the caller; the caller can then decide whether to log the result...and then there is no overlap or duplication.  Arguably, if the caller requests a retry and the retry succeeds, then from the caller's perspective, the operation was a success, and it's up to the callee to decide whether the initial failure is worth logging (e.g., we could defer logging the initial failure until after we know whether the retry succeeds...in this case, I don't think we want to do that, but I can imagine other circumstances in which that would be exactly what we want).

Optimally, perhaps we would have the callee do _no logging_ and instead return an "alternate success" status, which would enable the caller to log the result properly.  I doubt it's worth that level of complication, here, though.

webbnh commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1311,125 +1355,88 @@ def _send_directory(self, directory, uri, ctx):
-                else:
+                (_, tar_md5) = md5sum(tar_file)
+            except Exception:
+                self.logger.exception("Failed to read tar ball, '%s'", tar_file)
```
Yes!  What Dave said:  the exception will supply the "what happened" and/or "why it failed"; what we need the log message to provide is the context of what we were trying to do at the time.  So, yeah, the value of `tar_file` is key, and it would be good to mention that we had called `md5sum()` (although that might be apparent from the traceback).

dbutenhof commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -1,19 +1,40 @@
+        assert kwargs["stderr"] == subprocess.STDOUT
+        assert kwargs["stdout"] == subprocess.PIPE
+        c = subprocess.CompletedProcess(args, returncode=0, stdout=b"", stderr=None)
+        assert all(
+            x in c.args[0]
+            for x in [
+                "tar_path",
+                "--create",
+                "--xz",
+                "--force-local",
+                f"--file={tar_file}",
+            ]
+        )
```
It's a bit odd to be doing this assertion against your constructed `subprocess.CompletedProcess` rather than against the original argument. Functionally it's equivalent, but it inevitably and unnecessarily raises a momentary question in the reader's mind whether the method you called (even though it happens to be just a constructor) might have changed things.

On the other hand ... my first inclination was to also complain that you were merely checking that all expected arguments are in the list without regard to the order. Since all but one are options, though, there's no specific required order so  this is more flexible than `assert c.args == [ ... ]` as I might have suggested. (Although I really do hate what `black` does with long comprehensions. ☹️ )

npalaska replied:
> It's a bit odd to be doing this assertion against your constructed `subprocess.CompletedProcess` rather than against the original argument. Functionally it's equivalent, but it inevitably and unnecessarily raises a momentary question in the reader's mind whether the method you called (even though it happens to be just a constructor) might have changed things.

Yeah its the same thing, but I see how that might be confusing to readers thinking whether the method changes them. On the other hand I also wanted to make sure the arguments did not get modified unexpectedly by the subprocess function and all the original arguments passed to the subprocess are intact.

dbutenhof commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -57,37 +82,37 @@ class TestCreateTar:
     }
 
-    def test_create_tar(self, agent_setup, mock_tar):
+    def test_create_tar(self, agent_setup, mock_tar, caplog):
```
Do you need either the `caplog` fixture or the `self.logger.setLevel` here? This is a success case, right; so there shouldn't be any warnings?

dbutenhof commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -57,37 +82,37 @@ class TestCreateTar:
 
-    def test_create_tar_ignore_warnings(self, agent_setup, mock_tar_no_warnings):
+    def test_create_tar_ignore_warnings(
+        self, agent_setup, mock_tar_no_warnings, caplog
```
So does this mean that just bringing in the `caplog` fixture is enough to suppress pytest's display of the `WARNING` logs? Or are is this implicitly counting on inheriting the effect of the `self.logger.setLevel` from the previous test? (Which would be a bad idea.)

If you're sharing a logger across test cases, and using `setLevel`, be sure that you set it back. Or simply use `Logger.getLogger` in each test case where you need it, and just set that logger's level locally.

npalaska replied:
Yeah the logger was really taking effect across the test functions, so I created a logger fixture that just yields the logger by temporarily setting the log level and then set it back to `NOTSET` inside the fixture.

dbutenhof commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -57,37 +82,37 @@ class TestCreateTar:
+    def test_create_tar_ignore_warnings(
+        self, agent_setup, mock_tar_no_warnings, caplog
+    ):
         """Test if we can suppress the errors raised during the tar creation"""
```
The docstring isn't that useful, and you've repeated it in multiple test cases.

In any case, "suppress[ing] the errors" shouldn't really be the point of any of the unit tests; that's between you, your code, and pytest. What we want to actually test is the tool meister send's retry logic. Describing the purpose and strategy of each test is much more useful.

portante reviewed (changes requested):
I agree that the existing comments should be addressed.

dbutenhof commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,241 @@
+    def test_create_tar_failure(
+        self, agent_setup, mock_tar_failure, caplog, logger_fixture
+    ):
+        """Test if we can suppress the errors raised during the tar creation"""
```
Doesn't this actually verify that, without `retry`, we *report* the failure and a non-zero return status? We're not "suppressing" anything here.

dbutenhof commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,241 @@
+    def test_tar_create_success(self, agent_setup, monkeypatch, logger_fixture):
+        """This test should pass the tar creation in send directory"""
+
+        def fake_rmtree(directory: pathlib.Path):
+            return True
```
In a lot of mocks we've put together along the way, we add a mechanism to track whether the call was made. That is, if the call to `rmtree` is part of the method contract, we can use the mock to verify that. E.g.,

```
    rmtree_called = False

    def fake_rmtree(directory: pathlib.Path):
        rmtree_called = True
        return True
```

So that the main test case can then `assert rmtree_called`. (For extra points, we can pay attention to the `directory` parameter and assert that it was called with the proper path.)

But also ... [the documentation](https://docs.python.org/3/library/shutil.html#shutil.rmtree) doesn't specify a return value, only that exceptions are raised on error. Is the `return True` really valid? If the real `shutil.rmtree` doesn't `return True` then our mock shouldn't either.

npalaska replied:
Yeah I should use `pass` here and also nested functions can only access variables from outer functions, but are not allowed to modify them, we need to use something like `nonlocal` to reference that variable inside the inner function to change its value.
So the example would be
```
rmtree_called = False

def fake_rmtree(directory: pathlib.Path):
    nonlocal rmtree_called
    rmtree_called = True
    return True
```

dbutenhof commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,241 @@
+        monkeypatch.setattr("pbench.agent.tool_meister.md5sum", fake_md5)
+
+        def fake_unlink(*args):
+            return True
```
`path.unlink` doesn't return a value, either; it either raises an exception or returns `None`. You could just use `pass` here if you don't want to add something to assert that the call was made...

dbutenhof reviewed (approved):
I have some nits in the unit tests. I pointed out some places you could improve the tests with additional asserts. I'd prefer to avoid mocks with signatures distinctly different from the real code (e.g., your `return True` where there's no actual return value). This is all more a matter of code clarity (and possibly supportability) rather than functional correctness. (While I do worry a bit about someone reading the unit test and thinking "Oh, I can use `Path.unlink` as a truthy value in product code", it's not a high profile concern. 😆 )

portante commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,256 @@
+
+@pytest.fixture(scope="function")
+def logger_fixture(request):
+    logger = get_logger("__logger__")
```
Each logger is global state, so you should consider a unique name for the logger so that they are separaete.

npalaska replied:
The scope of this fixture is a function level so at the end of any individual test all the things defined inside the fixture should be gone isn't it? Why does the logger still remains there for whatever reason even after the fixture is gone

portante replied:
Because `get_logger` is from https://github.com/npalaska/pbench/blob/update_tar/lib/pbench/agent/tool_meister.py#L1712, which is just using `logging.getLogger()`, which is always global state (see https://docs.python.org/3.6/library/logging.html#logger-objects: "Multiple calls to `getLogger()` with the same name will always return a reference to the same `Logger` object.").

Also, it is not clear why the unit tests need to use the Tool Meister's `get_logger`.

portante reviewed (dismissed):
Looking good.

portante commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,256 @@
+
+@pytest.fixture(scope="function")
+def logger_fixture(request):
+    logger = logging.getLogger(f"__logger__{request.node.name}")
```
How does this make the logger unique for the test?

Perhaps just use a UUID?

portante replied:
So that is a bit obscure.  =)

Let's add a comment to make sure that assumption is understood clearly.

webbnh reviewed (approved):
This looks very good, especially the unit test for the `_send_directory()` cases.  I have a few small things for you to consider.  Also, there is one lingering conversation from my previous review about not using `agent_setup["tmp"]` if you can avoid it.

dbutenhof reviewed (approved):
I think the checking in the mocks could still be improved, but at this point I won't object. 🍤

dbutenhof commented on lib/pbench/agent/tool_meister.py:
```diff
@@ -1357,25 +1356,32 @@ def _send_directory(self, directory, uri, ctx):
+            self.logger.exception(
+                "Exception attempting to create the tarball, '%s'", tar_file
+            )
             failures += 1
```
Weird that I can't seem to locate this comment in the changes tab to reply as part of a review. Rebasing does weird things to GitHub.

Anyway, the point was that he's raising `ToolMeisterError` inside this try block, and wants it to propagate rather than stopping it here. This is one of the problems with blindly finalizing overly-broad exceptions classes like `Exception` ... but at the time the hack seemed reasonably expedient vs trying to work out a real solution. 😬

webbnh replied:
> Rebasing does weird things to GitHub.

Oh, it's even weirder than that (e.g., Nikhil has not yet rebased since I made the comment):  I started my review against only the last six commits, and ended it with comments against all 12, so if you're looking at the wrong set, you won't see my comment....  (And, for the record, the commit-range selector has been broken now for about 3 weeks...I need to report that...again.)  However, if you click on the [`lib/pbench/agent/tool_meister.py`](https://github.com/distributed-system-analysis/pbench/pull/2704/files/e260e6455d5a18e63e542c25b6bf195453102f19..9dc1f534e03403bc64b73e0bb4b845eb3e897496#diff-39405c6c2d5e89ba6fd6ba87a77fa9175dbdd28062af27e6d7c36f583bd0ba4b) link above the diffs above, you'll get there.  🙃 

> the point was that he's raising `ToolMeisterError` inside this try block, and wants it to propagate rather than stopping it here.

Yeah, I noted that, but I'm not clear on what that's accomplishing (as well as how the events leading up to the `ToolMeisterError` are any different from any other sort of exception-raising failure).

> This is one of the problems with blindly finalizing overly-broad exceptions classes like `Exception`

I noted that this scope is enclosed by a `try` block with a `finally` clause, and those are really resilient!  And, this one is prbably going to issue a warning when it cannot remove the tarball that we didn't create....

dbutenhof replied:
> However, if you click on the [...] link above the diffs above, you'll get there.

This is actually true, and it takes me to the *6 commits* you had selected at the time. Since the comment did not appear on *any* individual commit, apparently GitHub manages comments made against a multiple-select set of commits *separately* so that they can only be found in conversation view or against that specific set.

This is a bit mind blowing and surprisingly bizarre. How to go out of your way to be opaque and mysterious, GitHub!

webbnh replied:
I assume their thinking is that the comment relates to the specific differences being displayed, so displaying it for any other combination of commits might not make sense.  🦐

npalaska replied:
@webbnh Yes I do think handling `ToolMeisterError` differently than the broad `exception` is worthwhile, because if this exception is raised then we are certain that we failed to even create an empty tarball and it is better to propagate this error to higher level rather than logging it locally and continuing with our operation. There is no point of moving forwards if we encounter this error in this place of our TM workflow.

webbnh replied:
Looking at the code, it seems like the only difference between the handling of `ToolMeisterError` and any other `Exception` is that the later results in a normal return with a non-zero value while the former results in an exception propagation -- they both skip the rest of the normal processing code in `_send_directory()`, and they both execute the `finally` block.  The main difference occurs in the _callers_ of `_send_directory()` which look woefully unprepared to handle exceptions....

Nevertheless, if @dbutenhof is OK with this state of things, I shan't argue.

dbutenhof commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -168,46 +169,39 @@ def f(directory: pathlib.Path, tar_file: pathlib.Path):
                 args=[], returncode=returncode, stdout=stdout, stderr=None
             )
 
+        TestSendDirectory.function_called.append("fake_tar")
```
Yeah, this is one reason I really prefer using instance method test cases instead of static methods; you can just use `self.`. In any case, you could still use `__class__.` ... but while that's a bit more flexible (e.g., if the name were changes later), it's not all that much shorter.

npalaska replied:
Yeah but there are certain mocks like `open` or `unlink` that just take `*args` as a parameter and making them instance method is bit complex. If I do the following, then I can assert that the first argument is a tar path
```
@staticmethod
    def fake_open(*args):
        assert args[0] == pathlib.Path(f"{TestSendDirectory.directory}.tar.xz")
``` 
but if I do the following, args become this `('rb',) ` and there is no way to make assertion on the right tar file path
```
@staticmethod
    def fake_open(self, *args):
        # args == ('rb',) 
```

webbnh replied:
@npalaska, your last code blob there is internally inconsistent (at least by convention):  if it's a static method, then there should be no`self` parameter; but, if it's meant to be an instance method, then you need to remove the `@staticmethod` decorator.

By convention, the first parameter to an instance method should be named `self` (and the first parameter to a class method should be named `cls`), but Python doesn't actually care, and you can name it anything you want (but, you should follow the convention...because if you don't, your IDE and your reviewers are likely to get cranky 🙃).  When the function is called, what happens depends totally on what the caller **thinks** the function is:  if it thinks it is a static method, then the first parameter will be bound to the first argument; but, if the caller thinks it is an instance method, then the first parameter will be bound to a reference to the instance, and the second parameter will be bound to the first argument.

So, in last blob there, `self` gets bound to the `Path` instance, and the arguments are passed via `args`.  In your first blob, there's only one parameter, so all the arguments (including the instance ref) are passed in `args`.

So, it doesn't _really_ matter which one you use.  But, for sanity sake, it would be best if the function signature of your mock function properly matched the signature of the function that you're replacing.  In particular, if you are replacing an instance method, your mock should be an instance method -- with the first parameter named `self` and _no_ `@staticmethod` decorator -- and then I think things will follow naturally.

npalaska replied:
@webbnh sorry that was my mistake, thank you for pointing it out but the last blob of the code that I was testing did not have the `staticmethod` decorator it was just (I copied it from the first blob and forgot to remove it)
```
def fake_open(self, *args):
        # args == ('rb',) 
```
and as I said earlier I couldn't assert this statement `args[0] == pathlib.Path(f"{TestSendDirectory.directory}.tar.xz")`

webbnh replied:
Yes, that statement doesn't work because, when the mock function is an instance method, the path should be bound to `self` not to `args[0]`.

npalaska replied:
But the `path` is not bound to `self`, `self` is bound to `TestSendDirectory` object so the path is somehow lost.

npalaska replied:
These fixtures are now local to `test_tar_create_success`, so we are out of the woods now w.r.t this complexity.

webbnh reviewed (dismissed):
Looks quite good, but the signatures of the mock functions need some tweaking.

Also, I don't understand how the `md5sum()` mock is working.

npalaska commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -168,46 +169,39 @@ def f(directory: pathlib.Path, tar_file: pathlib.Path):
-            nonlocal function_called
-            function_called += 1
-            return 10, "random_md5"
+    @staticmethod
+    def fake_unlink(*args):
+        assert args[0] == pathlib.Path(f"{TestSendDirectory.directory}.tar.xz")
+        TestSendDirectory.function_called.append("fake_unlink")
+        pass
```
`self` is not pathlib.Path object, it is an instance of `TestSendDirectory` class, so this assertion fails `assert self == pathlib.Path(f"{TestSendDirectory.directory}.tar.xz")`

npalaska commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -168,46 +169,39 @@ def f(directory: pathlib.Path, tar_file: pathlib.Path):
+        pass
 
-        monkeypatch.setattr(pathlib.Path, "unlink", fake_unlink)
+    def fake_md5(self, tar_file: pathlib.Path):
+        assert tar_file == pathlib.Path(f"{self.directory}.tar.xz")
+        self.function_called.append("fake_md5")
+        return 10, "random_md5"
```
I believe the following hypothesis is happening:
As I mentioned above `self` is bound to a `TestSendDirectory` object, when the source code calls the `md5sum` function, the test refers to whatever is patched by the `monkeypatch` (in this case its a `fake_md5` instance method of `TestSendDirectory` object), this instance method also takes the same arguments as the original method so that works.

portante replied:
See if https://github.com/npalaska/pbench/pull/2 addresses this satisfactorily.

portante reviewed (commented):
Can we consider suggestions in https://github.com/npalaska/pbench/pull/2?

portante commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -168,46 +169,39 @@ def f(directory: pathlib.Path, tar_file: pathlib.Path):
+    def fake_rmtree(self, directory: pathlib.Path):
+        assert directory == tmp_dir
+        self.function_called.append("fake_rmtree")
+        pass
```
See if https://github.com/npalaska/pbench/pull/2 addresses this satisfactorily.

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,202 @@
+
+        with pytest.raises(ToolMeisterError) as exc:
+            failures = tool_meister._send_directory(self.directory, "uri", "ctx")
+            assert failures == 1
```
Are you sure you want this `assert` statement inside the `with` block?

portante replied:
Good catch, that is bad.

webbnh replied:
I don't know that it is bad, per se -- if the assertion fails, the test will certainly fail -- but it's probably not good.  🙂

portante replied:
That assertion never fires so it was removed.

webbnh replied:
Oh, yes...of course.  (I.e., "Well, duh!"  😆)

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,202 @@
+                args=[], returncode=returncode, stdout=stdout, stderr=None
+            )
+
+        __class__.functions_called.append("fake_create_tar")
```
Uh...eh-hem...shouldn't this call to `append()` be inside the mock function (i.e., before line 134)??  😆

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,202 @@
+            assert failures == 1
+        functions_called, self.functions_called = self.functions_called, []
+        assert functions_called == ["fake_create_tar"]
+        assert f"Failed to create an empty tar {self.directory}.tar.xz" in str(
+            exc.value
```
Once the mock for `_create_tar()` is fixed to actually record the calls to the function, I don't think that this assertion will be necessary (that is, from previous assertions, we'll already know that there were two calls to `_create_tar()` and that the `ToolMeisterError` was raised, so we don't need to also confirm the contents of the log, which we probably shouldn't be doing in this case, anyway).

portante replied:
Contents of the log or the exception?

webbnh replied:
Whoops!  My bad...I was thinking that we were looking at the log output, not the exception message.

I'm not sure I care whether we test the exception message or not (especially given that we've already check that we received the exception).

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,202 @@
+
+    def test_tar_create_failure(self, tool_meister, monkeypatch):
+        """Check if the tar creation error is properly captured in send_directory"""
+        monkeypatch.setattr(
+            tool_meister,
+            "_create_tar",
+            self.fake_create_tar(1, b"Error in tarball creation"),
+        )
```
We should also be mocking `Path.unlink()` in this test, since it will be called from the `finally` block when `ToolMeisterError` is raised.

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,202 @@
+"""
+import io
+import logging
+import pathlib
```
Consider using `from pathlib import Path` instead.  (This is the overwhelmingly typical form in our code, and it eases subsequent references to `Path`.)

webbnh reviewed (changes requested):
Most of my comments below are just suggestions.  However, I did find a bug that I think you should address, and so I'm hoping that you'll take me up on my suggestions, as well.

For your especial consideration:
- The mock produced by `fake_create_tar()` should be adding to the `functions_called` list, and it's _not_.
- While you're fixing that, I strongly suggest changing `functions_called` from a class attribute to a local variable (which I believe is very easy to do), which will relieve us from having manage its value/lifetime.
- Also, the mock functions which replace instance methods should themselves be defined as instance methods.  (This shouldn't be hard to do.)
- `test_tar_create_failure()` should set up a mock for `Path.unlink()`, since it will be called from the finally block when `ToolMeisterError` is raised.
- And, just as an aside, I prefer the term "mock" over "fake".

portante commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -15,241 +15,188 @@
 
 class TestSendDirectory:
-    """Test create_tar in send directory of tool meister"""
+    """Test ToolMeister._send_directory()'s use of ._create_tar()"""
```
Can we make that change when that happens instead?

webbnh replied:
Yeah, we could...but, I'm hoping that, since Nikhil has another change to make anyway, he could do it now.

portante commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -139,60 +135,47 @@ def f(directory: pathlib.Path, tar_file: pathlib.Path):
-        failures = tm._send_directory(self.directory, "uri", "ctx")
-        assert self.function_called == [
+        failures = tool_meister._send_directory(self.directory, "uri", "ctx")
+        functions_called, self.functions_called = self.functions_called, []
```
Use of `nonlocal` makes sense.

webbnh replied:
I don't know that `nonlocal` will "reach".  (I know that Python has "interesting" scoping rules, but I'm not sure that the invocation of a static class method can reference a local variable in its caller's scope with a simple `nonlocal` definition, because, lexically, the caller is not a "containing" scope...but, I don't know what the dynamic scoping rules are.)

On the flip side, though, I don't think it's a problem for the inner function of `fake_create_tar()` to access the outer function's parameters, and Python provides "pass by name" semantics for parameters, so the mock could have access to the local list without the use of `nonlocal` if we just pass it in and let it be part of the closure.

portante replied:
I spoke with @npalaska, this works just fine with a local variable, `nonlocal` is not needed.

portante commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -0,0 +1,202 @@
+            failures = tool_meister._send_directory(self.directory, "uri", "ctx")
+            assert failures == 1
+        functions_called, self.functions_called = self.functions_called, []
+        assert functions_called == ["fake_create_tar"]
```
It is not because of how the mock works.  That is a problem.

portante reviewed (changes requested):
I am partly responsible for the test changes so let me know if you need anything to make those changes.

npalaska commented:
> I am partly responsible for the test changes so let me know if you need anything to make those changes.

Thanks Peter, I am looking into addressing these suggestions right now, I will push the changes shortly

portante commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -139,60 +135,47 @@ def f(directory: pathlib.Path, tar_file: pathlib.Path):
+        def fake_open(*args):
+            assert args[0] == pathlib.Path(f"{self.directory}.tar.xz")
+            self.functions_called.append("fake_open")
+            return io.StringIO()
```
Why do we need to do this?  In Python a function invocation is the same.  The mock'd function just needs to verify what it is expected to be called with, so this is just syntactic sugar, no?

webbnh replied:
I was prepared to embrace your argument until earlier when Nikhil suggested having evidence that Python's calling conventions are not as simple as your prompt here suggests.

Also, I think it makes it easier for the reader if the thing being called closely parallels what the caller expects it to be.  (And, it's not hard to wrap these two functions in an inner class.)

The better question is, should we be refactoring the beautiful `MockedPath` class that you (@portante) have implemented to make it available for use here?

npalaska replied:
Hmm, I see that @portante has implemented the `MockedPath` class in his PR #2725, I could implement something similar here (which will eventually gets merged/rebased on Peter's changes anyway) or I could keep it as it is for now and let peter add these in his `MockedPath` class when doing the rebase :upside_down_face:

webbnh replied:
I would vote for implementing something similar here now, rather than counting on the future work to remember to address this.  I think what is needed for this PR is pretty simple.

portante replied:
I have to do a bit of surgery to merge with PR #2725, so I can add that there and track the work during the integration.

webbnh replied:
Okie....

webbnh commented on lib/pbench/test/unit/agent/test_tool_meister.py:
```diff
@@ -67,105 +67,111 @@ def fake_run(*args, **kwargs):
-            assert args[0] == pathlib.Path(f"{self.directory}.tar.xz")
-            self.functions_called.append("fake_open")
+        def mock_open(*args):
+            assert args[0] == Path(f"{TestSendDirectory.directory}.tar.xz")
```
Oddly enough, I believe that we can use `self` in this context to access `directory`.  (E.g., see line 163 in `mock_md5()`. 🙂)  Ditto lines 150 and 199.

npalaska replied:
Yep!

webbnh reviewed (approved):
It looks great!

(Just one small thing (which appears a few times...)...which we can ignore if we need to....)

webbnh reviewed (approved):
🎉

npalaska commented:
PBENCH-83 #status closed #comment trying to close the ticket via github

npalaska commented:
PBENCH-83 #comment trying to close the ticket via github
